library/ssl_srv.c

```diff
@@ -3879,12 +3863,13 @@ static int ssl_decrypt_encrypted_pms( mbedtls_ssl_context *ssl,
     defined(MBEDTLS_SSL_PROTO_TLS1_2)
     if( ssl->minor_ver != MBEDTLS_SSL_MINOR_VERSION_0 )
     {
-        if ( p + 2 > end ) {
+        if ( p + 2 > end )
+        {
             MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad client key exchange message" ) );
             return( MBEDTLS_ERR_SSL_BAD_HS_CLIENT_KEY_EXCHANGE );
         }
-        if( *p++ != ( ( len >> 8 ) & 0xFF ) ||
-            *p++ != ( ( len      ) & 0xFF ) )
+        if( *p++ != MBEDTLS_BYTE_1( len ) ||
```

Comment:
MBEDTLS_GET_UINT16_BE could be used here.

Reply:
I couldn't actually get this to work here
If I did `MBEDTLS_GET_UINT16_BE( len, 0 )` it says that the macro requires a pointer-to-object as it tries to index it. I think we should leave this as retrieving the bytes as it cannot be indexed
```
  CC    ssl_srv.c
In file included from ssl_srv.c:20:
ssl_srv.c: In function ‘ssl_decrypt_encrypted_pms’:
common.h:207:32: error: subscripted value is neither array nor pointer nor vector
  207 |           ( (uint16_t) ( data )[( offset )    ] << 8 )          \
      |                                ^
ssl_srv.c:3467:14: note: in expansion of macro ‘MBEDTLS_GET_UINT16_BE’
 3467 |     if( p != MBEDTLS_GET_UINT16_BE( len, 0 ) )
      |              ^~~~~~~~~~~~~~~~~~~~~
common.h:208:32: error: subscripted value is neither array nor pointer nor vector
  208 |         | ( (uint16_t) ( data )[( offset ) + 1]      )          \
      |                                ^
ssl_srv.c:3467:14: note: in expansion of macro ‘MBEDTLS_GET_UINT16_BE’
 3467 |     if( p != MBEDTLS_GET_UINT16_BE( len, 0 ) )
      |              ^~~~~~~~~~~~~~~~~~~~~
make[1]: *** [Makefile:271: ssl_srv.o] Error 1
make: *** [Makefile:18: lib] Error 2
```
![image](https://user-images.githubusercontent.com/85489920/130435299-3dff2fd9-a806-4f74-87da-b45dfd3325fc.png)

Reply:
And if I compare the length to the first to bytes of p
`if len != MBEDTLS_GET_UINT16_BE( p, 0 )` it flags a warning, as you are not comparing to variables of the same data type

Reply:
You would need to case `len` to an uint16_t or MBEDTLS_GET_... to `size_t` but never mind.